File: 0067-add-binary/67.add-binary.cs
/*
 * @lc app=leetcode id=67 lang=csharp
 *
 * [67] Add Binary
 */

// @lc code=start
public class Solution {
    public string AddBinary(string a, string b) {
        List<char> result = new List<char>();
        int i = a.Length - 1;
        int j = b.Length - 1;
        int carry = 0;

        while (i >= 0 || j >= 0) {
            int sum = carry;

            if (i >= 0) {
                sum += a[i] - '0';
                i--;
            }
            if (j >= 0) {
                sum += b[j] - '0';
                j--;
            }

            result.Add((char)(sum % 2 + '0'));
            carry = sum / 2;
        }

        if (carry != 0) {
            result.Add((char)(carry + '0'));
        }
        
        result.Reverse();

        return new string(result.ToArray());
    }
}
// @lc code=end

File: 0067-add-binary/67.add-binary.rs
/*
 * @lc app=leetcode id=67 lang=rust
 *
 * [67] Add Binary
 */

// @lc code=start
impl Solution {
    pub fn add_binary(a: String, b: String) -> String {
        let mut a = a.as_bytes();
        let mut b = b.as_bytes();
        let mut carry = 0;
        let mut result = Vec::new();

        while !a.is_empty() || !b.is_empty() || carry != 0 {
            let mut sum = carry;

            if let Some(&x) = a.last() {
                sum += x - b'0';
                a = &a[..a.len() - 1];
            }
            if let Some(&x) = b.last() {
                sum += x - b'0';
                b = &b[..b.len() - 1];
            }

            carry = sum / 2;
            result.push((sum % 2 + b'0') as char);
        }

        result.into_iter().rev().collect()
    }
}
// @lc code=end

